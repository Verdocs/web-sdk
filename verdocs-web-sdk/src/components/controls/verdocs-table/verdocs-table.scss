@import '../../../theme.scss';

verdocs-table {
  font-family: $primary-font;
  box-sizing: border-box;
  margin: 0 0 10px 0;
  display: block;
  width: 100%;

  .table-wrapper {
    margin: 10px 0;
    padding: 0;
  }

  .table {
    width: 100%;
    border: none;
  }

  .row {
    &.header-row {
    }

    &.data-row {
    }
  }

  .header-col {
    text-align: left;
    padding: 3px 8px;
    font-weight: 700;
    color: #555555;
  }

  .data-col {
    padding: 3px 8px;
    font-weight: 400;
  }
}
