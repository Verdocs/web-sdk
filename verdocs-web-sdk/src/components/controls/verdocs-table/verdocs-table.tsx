import {Component, Prop, Host, h, Event, EventEmitter} from '@stencil/core';

export interface IColumn {
  id: string;
  header?: string;
  renderHeader?: (col: IColumn) => any;
  renderCell?: (col: IColumn, row: any) => any;
}

/**
 * Display a simple table of data. Columns and data cells may have custom renderers defined to
 * support creating interactive table layouts.
 */
@Component({
  tag: 'verdocs-table',
  styleUrl: 'verdocs-table.scss',
})
export class VerdocsTable {
  /**
   * The columns to display
   */
  @Prop() columns: IColumn[] = [];

  /**
   * The rows to display
   */
  @Prop() data: any[] = [];

  /**
   * Event fired when the user clicks a column header. This may be used to manage sorting options.
   */
  @Event({composed: true}) colHeaderClick: EventEmitter<{col: IColumn}>;

  /**
   * Event fired when the user clicks a row.
   */
  @Event({composed: true}) rowClick: EventEmitter<{row: any}>;

  render() {
    return (
      <Host class="table-wrapper">
        <table class="table">
          <thead>
            <tr class="row header-row">
              {this.columns.map((col, i) => {
                return (
                  <th
                    class={`col header-col col-${i} col-${col.id}`}
                    onClick={() => this.colHeaderClick?.emit({col})}
                    innerHTML={col.renderHeader ? col.renderHeader(col) : col.header || col.id}
                  />
                );
              })}
            </tr>
          </thead>

          <tbody>
            {this.data.map(row => (
              <tr class="row data-row" onClick={() => this.rowClick?.emit({row})}>
                {this.columns.map((col, i) => {
                  return <td class={`col data-col col-${i} col-${col.id}`}>{col.renderCell?.(col, row) || row[col.id]}</td>;
                  // return <td class={`col data-col col-${i} col-${col.id}`} innerHTML={col.renderCell?.(col, row) || row[col.id]} />;
                })}
              </tr>
            ))}
          </tbody>
        </table>
      </Host>
    );
  }
}
