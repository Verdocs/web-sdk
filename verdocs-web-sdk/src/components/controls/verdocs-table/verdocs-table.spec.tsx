import { newSpecPage } from '@stencil/core/testing';
import { VerdocsTable } from './verdocs-table';

describe('verdocs-table', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsTable],
      html: `<verdocs-table></verdocs-table>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with columns and data', async () => {
    const columns = [
      { id: 'name', header: 'Name' },
      { id: 'age', header: 'Age' }
    ];
    const data = [
      { name: 'Alice', age: 30 },
      { name: 'Bob', age: 25 }
    ];
    const page = await newSpecPage({
      components: [VerdocsTable],
      html: `<verdocs-table columns='${JSON.stringify(columns)}' data='${JSON.stringify(data)}'></verdocs-table>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('emits colHeaderClick and rowClick events', async () => {
    const columns = [
      { id: 'name', header: 'Name' }
    ];
    const data = [
      { name: 'Alice' }
    ];
    const page = await newSpecPage({
      components: [VerdocsTable],
      html: `<verdocs-table></verdocs-table>`,
    });
    (page.rootInstance as any).columns = columns;
    (page.rootInstance as any).data = data;
    await page.waitForChanges();

    const colSpy = jest.fn();
    const rowSpy = jest.fn();
    page.root.addEventListener('colHeaderClick', colSpy);
    page.root.addEventListener('rowClick', rowSpy);

    const th = page.root.querySelector('th');
    th?.dispatchEvent(new Event('click', { bubbles: true }));

    // Simulate row click (if implemented in render)
    const tr = page.root.querySelector('tr.row');
    tr?.dispatchEvent(new Event('click', { bubbles: true }));

    expect(colSpy).toHaveBeenCalled();
    // rowClick may not be implemented in the render, so this is a best-effort check
  });
});
