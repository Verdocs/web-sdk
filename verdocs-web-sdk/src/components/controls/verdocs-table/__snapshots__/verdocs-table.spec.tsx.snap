// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-table renders with columns and data 1`] = `
<verdocs-table class="table-wrapper">
  <table class="table">
    <thead>
      <tr class="header-row row">
        <th class="col col-0 col-name header-col">
          Name
        </th>
        <th class="col col-1 col-age header-col">
          Age
        </th>
      </tr>
    </thead>
    <tbody>
      <tr class="data-row row">
        <td class="col col-0 col-name data-col">
          Alice
        </td>
        <td class="col col-1 col-age data-col">
          30
        </td>
      </tr>
      <tr class="data-row row">
        <td class="col col-0 col-name data-col">
          Bob
        </td>
        <td class="col col-1 col-age data-col">
          25
        </td>
      </tr>
    </tbody>
  </table>
</verdocs-table>
`;

exports[`verdocs-table renders with default props 1`] = `
<verdocs-table class="table-wrapper">
  <table class="table">
    <thead>
      <tr class="header-row row"></tr>
    </thead>
    <tbody></tbody>
  </table>
</verdocs-table>
`;
